include/madwf_ml.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <vector>
+#include <random>
+#include <unordered_map>
+
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <polynomial.h>
+
+#include <random_quda.h>
+#include <tune_quda.h>
+
+#include <device_vector.h>
+#include <invert_quda.h>
+#include <madwf_param.h>
+
+namespace quda
+{
+  namespace madwf_ml
+  {
+    /**
+      @brief defines the types of 5th dimension transfer matrices:
+      - Wilson: has both the color and spin indices, d.o.f = (4 x 3) * (4 x 3) = 144
+      - Spin: only has the spin index, d.o.f = (4) * (4) = 16
+      - Chiral: only has the chiral index, d.o.f = 2
+    */
+    enum class transfer_5D_t { Wilson = 144, Spin = 16, Chiral = 2 };
+
+    /**
+      @brief A helper class to instantiate the precisions (half and quarter) for madwf_tensor and madwf_transfer
+    */
+    template <template <class T> class F, class... Args>
+    void instantiate_madwf(ColorSpinorField &out, const ColorSpinorField &in, Args... args)
```

Comment:
Just noting that now we are C++17, so we could write this without the preprocessor using `if constexpr` (I plan to make this change to instantiate.h at some point.

Reply:
https://github.com/lattice/quda/pull/1061/commits/703f9d363082c78607d77ac0d576e66733264f35

Reply:
Can we use `instantiatePreconditioner` for this?